chunk/chunk.go

```diff
@@ -5,124 +5,252 @@ import (
+	ErrInvalidChecksum = errors.Error("invalid chunk checksum")
+	ErrWrongMetadata   = errors.Error("wrong chunk metadata")
+)
+
+var castagnoliTable = crc32.MakeTable(crc32.Castagnoli)
+
 // Chunk contains encoded timeseries data
 type Chunk struct {
-	ID      string       `json:"-"`
+	// These two fields will be missing from older chunks (as will the hash).
+	// On fetch we will initialise these fields from the DynamoDB key.
+	Fingerprint model.Fingerprint `json:"fingerprint"`
+	UserID      string            `json:"userID"`
+
+	// These fields will be in all chunks, including old ones.
 	From    model.Time   `json:"from"`
 	Through model.Time   `json:"through"`
 	Metric  model.Metric `json:"metric"`
 
+	// The hash is not written to the external storage either.  We use
+	// crc32, Castagnoli table.  See http://www.evanjones.ca/crc32c.html.
+	// For old chunks, ChecksumSet will be false.
+	ChecksumSet bool   `json:"-"`
+	Checksum    uint32 `json:"-"`
+
 	// We never use Delta encoding (the zero value), so if this entry is
 	// missing, we default to DoubleDelta.
 	Encoding prom_chunk.Encoding `json:"encoding"`
 	Data     prom_chunk.Chunk    `json:"-"`
 
+	// This flag is used for very old chunks, where the metadata is read out
+	// of the index.
 	metadataInIndex bool
 }
 
 // NewChunk creates a new chunk
-func NewChunk(fp model.Fingerprint, metric model.Metric, c prom_chunk.Chunk, firstTime, lastTime model.Time) Chunk {
+func NewChunk(userID string, fp model.Fingerprint, metric model.Metric, c prom_chunk.Chunk, from, through model.Time) Chunk {
 	return Chunk{
-		ID:       fmt.Sprintf("%d:%d:%d", fp, firstTime, lastTime),
-		From:     firstTime,
-		Through:  lastTime,
-		Metric:   metric,
-		Encoding: c.Encoding(),
-		Data:     c,
+		Fingerprint: fp,
+		UserID:      userID,
+		From:        from,
+		Through:     through,
+		Metric:      metric,
+		Encoding:    c.Encoding(),
+		Data:        c,
 	}
 }
 
-func parseChunkID(id string) (model.Fingerprint, model.Time, model.Time, error) {
-	parts := strings.Split(id, ":")
+// parseExternalKey is used to construct a partially-populated chunk from the
+// key in DynamoDB.  This chunk can then be used to calculate the key needed
+// to fetch the Chunk data from Memcache/S3, and then fully populate the chunk
+// with decode().
+//
+// Pre-checksums, the keys written to DynamoDB looked like
+// `<fingerprint>:<start time>:<end time>` (aka the ID), and the key for
+// memcache and S3 was `<user id>/<fingerprint>:<start time>:<end time>.
+// Finger prints and times were written in base-10.
+//
+// Post-checksums, externals keys become the same across DynamoDB, Memcache
+// and S3.  Numbers become hex encoded.  Keys look like:
+// `<user id>/<fingerprint>:<start time>:<end time>:<checksum>`.
+func parseExternalKey(userID, externalKey string) (Chunk, error) {
+	if !strings.Contains(externalKey, "/") {
+		return parseLegacyChunkID(userID, externalKey)
+	}
+	return parseNewExternalKey(externalKey)
```

Comment:
If that happens, the chunk will fail at the decode phase.  The `userID` is only used for the legacy functions.